types/index.d.ts

```diff
@@ -9,7 +9,11 @@ declare function dayjs (date?: dayjs.ConfigType, format?: dayjs.OptionType, stri
 declare function dayjs (date?: dayjs.ConfigType, format?: dayjs.OptionType, locale?: string, strict?: boolean): dayjs.Dayjs
 
 declare namespace dayjs {
-  export type ConfigType = string | number | Date | Dayjs
+  interface ConfigTypeMap {
+    default: string | number | Date | Dayjs
+  }
+
+  export type ConfigType = ConfigTypeMap[keyof ConfigTypeMap]
```

Comment:
what's this used for please? And will this affect the original type if the plugin is imported?

Reply:
@iamkun The `type` keyword doesn't support [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html). And yes, it will add the plugin's type into `ConfigType`.

`ConfigTypeMap[keyof ConfigTypeMap]` basically means "merge every members' type into a union". If no plugin was imported, the type of `ConfigType` is `string | number | Date | Dayjs` (same as before). When the ArraySupport plugin was imported, `ConfigType` will become `string | number | Date | Dayjs | [year?: number, ...]`.

Reply:
looks cool, does this requires higher ts version?

Reply:
@iamkun Labeled tuple elements require TypeScript 4.0 or higher, I can convert it back to un-labelled ones if you want to support older TypeScript versions.

Reply:
that's what I'm concerned. We should have a fallback support at least ts 3 I think

Reply:
@iamkun It should work on TypeScript 3 now.

Reply:
can we make sure that `ConfigTypeMap[keyof ConfigTypeMap]` works on ts3?

Reply:
> can we make sure that `ConfigTypeMap[keyof ConfigTypeMap]` works on ts3?

@iamkun Yes, this will work on TypeScript 3. You can verify it on [TypeScript playground](https://www.typescriptlang.org/play?ts=3.3.3#code/JYOwLgpgTgZghgYwgAgMpwLYAcA2ECycWAUAN7HKXIAmE8ArjmAFzIDOYUoA5sgD7IQ9DACNoAbgpUA9mAAW0VgBE4kSQF9ixMAE8sKdNjzIAvGky4CRANoBrCDukxzRq1gC6k4kA).